pkg/io/v2/opener.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v2
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+	"gocloud.dev/blob"
+	"gocloud.dev/gcerrors"
+
+	"k8s.io/test-infra/pkg/io/v2/providers"
+)
+
+// Aliases to types in the standard library
+type (
+	ReadCloser  = io.ReadCloser
+	WriteCloser = io.WriteCloser
+)
+
+// Opener has methods to retrieve a Reader or Writer for the given path
+type Opener interface {
+	Reader(ctx context.Context, path string, opts *blob.ReaderOptions) (io.ReadCloser, error)
+	Writer(ctx context.Context, path string, opts *blob.WriterOptions) (io.WriteCloser, error)
+}
+
+type opener struct {
```

Comment:
I would expect to be able to pass both gcs and aws credentials and be able to read from/write to both locations.

Reply:
I could implement is the only question is in which way:

* preserve the original file formats and change the signature to `func NewOpener(gcsCredentialsFile, awsCredentialsFile string) (Opener, error) {` I think this also means we need additional flags for each credentialsfile (even more if we would start supporting Azure,...)
* Only use one file but change the JSON format, e.g.:
```
{
  "s3_credentials": {
    "region": "us-east-1",
    ...
  },
  "gs_credentials": {
    "type": "service_account",
    ...
  }
}
```

Reply:
@fejta why? Do we have any code that does this? And wouldn't it be more logical to then construct two openers?

If possible I would like to keep this PR scoped to replicating the functionality in `pkg/io` and not add new features. If we want that, we can still do that in a follow-up.

Reply:
I should be able to have `this` repo write to S3 and `that` one write to GCS. Right now that is not possible because we do not support S3. It will still not be possible after this change because we still only support a single provider.

The point of this abstraction is that we can pass crier, tide, sidecar, etc an interface that allows it to remain ignorant of whether it is talking to the local filesystem, a PVC, or some remote storage system.

I would also like to keep the credentials flags specific to the provider, aka keep `--gcs-credentials` and `--aws-credentials`. GCP credentials are specific to GCP. We should not risk leaking them to AWS (and visa versa).

Reply:
Okay so this variant would mean we have to add a flag for every provider we want to support via explicit configuration via a config file (auto-discovery could still be supported for all providers via gocloud). Then I also should remove the blob-storage-credentials-file flag and use the other flag only to decide if the old or new package is used. Obviously I would remove the deprecation of the gcs-credentials-file flag.

Reply:
Okay. So how do we get consensus on this issue? :) For me it's okay either way

Reply:
Yeah, Erick has a point, someone might want to multiple providers

Reply:
Should be fixed now